### ISSUE TITLE
Clarification on workerCacheLookupStart

### ISSUE THREAD

**Issue by ErikWitt** [WICG/service-worker-static-routing-api#35]
Hey! The explainer looks very interesting. Especially `workerMatchedRouterSource` is very useful, we would certainly track it. `workerFinalRouterSource` is very cool as well as soon as `race-network-and-fetch-handler` is fixed (see [crbug](https://issues.chromium.org/issues/351796122)). 

I am not sure we will be able to test the OT because it is hard to get our customers to add the token as header to the service worker. A way for 3rd party origin trials for service worker features would be awesome in the future :) (our script is imported into the customer hosted service worker file).

**What I do not quite get from the explainer**, will I get workerCacheLookupStart even when I do the cache lookup manually from the fetchEvent or only if I specify cache as a static route?

PS: we would also be interested in `workerRouterEvaluationStart`. We are using static routing already a lot and are seeing very good performance gains in RUM (up to 300ms in median for returning users on android). Understanding the performance of static routes even deeper, could be valuable.

**Comment by quasi-mod** [comment]
Thanks for the feedback! 

> I am not sure we will be able to test the OT because it is hard to get our customers to add the token as header to the service worker. A way for 3rd party origin trials for service worker features would be awesome in the future :) (our script is imported into the customer hosted service worker file).

Let me check and see what we can do to enable 3rd parties.

> What I do not quite get from the explainer, will I get workerCacheLookupStart even when I do the cache lookup manually from the fetchEvent or only if I specify cache as a static route?

Sorry for the confusion. Current OT only shows `workerCacheLookupStart` for cache lookup with static routing API. 
We want to expand the `workerCacheLookupStart` for fetchEvent  in general (access to cache-storage in general), but since it requires to address quite a lot of challenges (e.g. what to do if we have multiple cache lookups in a fetch), generalization is still a work in progress.

> PS: we would also be interested in workerRouterEvaluationStart. We are using static routing already a lot and are seeing very good performance gains in RUM (up to 300ms in median for returning users on android). Understanding the performance of static routes even deeper, could be valuable.

Thank you so much! This is one of the things we really wanted to hear back in this OT.
As of this OT, we implemented the field and can be tested as well.
For the spec part, we will discuss them and get back to you.

**Comment by quasi-mod** [comment]
CC: @sisidovski @yoshisatoyanagisawa

**Comment by yoshisatoyanagisawa** [comment]
Thanks for the feedback and question :)

For the 3rd party OT, we have recognized the issue (https://crbug.com/40278041), but we could not prioritize it for a while.  I also remember is is not trivial to implement.

For `workerCacheLookupStart`, as Keita mentioned, it has been limited to the SW static routing API.  There was a suggestion to generalize it to the fetch handler, but we have not implemented yet because the semantics on the field is not clear.  Especially, the case that there is multiple cache look ups or cache look up won't be used in the `respondWith()`.  We would like to have feedbacks on this.

And thanks for the feedback on `workerRouterEvaluationStart`.  There is also a discussion on the field necessity.  If there is more supporting materials how it is useful for you, it helps us to push the field necessity in the WG.

**Comment by sisidovski** [comment]
@yoshisatoyanagisawa This origin trial is used to extend the resource timing API. It exposes the service worker related fields, but the enabled feature does not belong to the ServiceWorker itself. Unlike the [previous OT](https://developer.chrome.com/blog/service-worker-static-routing-api-origin-trial#trying_out_the_service_worker_static_routing_api) that we held for the static routing, I assume this OT doesn't have restrictions from ServiceWorker. So 3rd party OT should be available, but probably we have to update or recreate the entry in the OT dashboard. https://developer.chrome.com/origintrials/#/view_trial/1689412810217357313

**Comment by yoshisatoyanagisawa** [comment]
@sisidovski Yeah, that is true.  3rd party should also be able to run the OT.  And as you touched and I also understood from https://developer.chrome.com/docs/web-platform/third-party-origin-trials, there might need to have the option to request a third-party token.

**Comment by quasi-mod** [comment]
@ErikWitt Hi! 
This is just a FYI, but we will be enabling 3rd party OTs for this feature from Chrome 133 (will be on stable channel at the end of January). 
We will let you know when its live. You will need to update the OT token (re-register) once it is ready. Thanks!

**Comment by quasi-mod** [comment]
@ErikWitt Hi. Since Chrome 133 has been released, you can now use the 3rd Party OTs.

Please re-register to the OT with the "Third-Party matching option" enabled. 
Note that third party OTs will only work with Chrome 133 and later, and will not work with previous releases.

Let us know if there is anything we can do.
Thanks!

**Comment by dertheodor** [comment]
Hey @quasi-mod , 
I'm Theo, a colleague of Erik.

I was just testing out the feature and have re-registered the OT beforehand (with 3rd party and match subdomains set to both to yes). I couldn't find the exposed `window.performance.timing` data.
I created a small site to show you this https://theotest.app.baqend.com/v1/file/www/index.html. Maybe I am missing out on something. 

Would be grateful for any kind of help! :)

**Comment by quasi-mod** [comment]
Hi @dertheodor !

Thanks for testing the OT!
Let me clarify so that I am on the same page with you.

Looking at what you gave me, it seems like you are using the old [PerformanceTiming API](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceTiming) (accessed with `window.performance.timing`). 

Our feature is implemented on the new PerformanceResourceTimingAPI and [PerformanceNavigationTiming API](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceNavigationTiming).

From my local testing, I was able to see the new fields using `performance.getEntriesByType('resource')` and `performance.getEntriesByType('navigation')`. Could you confirm if you could see the new fields with them?

Thanks so much!

**Comment by dertheodor** [comment]
Hi @quasi-mod, thanks for the clarification! That works.
I was referring to the MD from here [resource-timing-api.md](https://github.com/WICG/service-worker-static-routing-api/blob/main/resource-timing-api.md#network-rule-specified) where the code was showing usage of the old `window.performance.timing` API :) 
We will implement this soon and hopefully be back with some more feedback . Greetings!

**Comment by quasi-mod** [comment]
@dertheodor Thanks for confirming, and extremely sorry about the wrong example.
I've updated the examples on the explainer.

Looking forward to hearing back from you all :).
Thanks!

**Comment by ErikWitt** [comment]
To get a better understanding of TTFB changes, I check out the attribution data before static routing vs. the time after we implemented the new tracking and static routing.

The chart shows
* Time to workerStart  reduced quite a bit (or has changed to `workerRouterEvaluationStart`)
* Service Worker init reduced
* DNS and TCP/TLS become visible (with Service Worker they were hidden in the responseTime I guess)
* Server response time has reduced (in part because DNS and connection are seperated but even when I add dns and connecting back to responseTime, it is still 100ms faster than before)

Since this is a before vs. after comparison, I cannot 100% be sure that there was a TTFB improvement incidentally at the same point in time, I think it is rather unlikely.

What I do not quite understand is
* Why did bother the time to `workerStart` as well as `swInit`  reduce? Or rather, what is captured as the time before workerStart and what is captured by service worker init exactly?
* Why did the response time improve as well? I would have expected the improved to come only from more efficient service worker handling

![Image](https://github.com/user-attachments/assets/b347c3ba-3b48-47cf-bce1-241627db55b5)

**Comment by quasi-mod** [comment]
Hi @ErikWitt , thanks for the feedback, and sorry for the delay.
The feedback you gave us is really interesting, so let me ask a couple of questions so that we can understand more.

- Is there any site or demo we could use to locally reproduce your problems?
- What kind / how many rules are specified in the router?
- How are each fields (`workerRouterEvaluationStart`, `workerStart`, `SW Init`, etc) calculated?
- Just to confirm, all of the data here is taken from the Performance Navigation Timing API (i.e. main resource)?

Generally speaking, the router evaluation time should be in microsecond order, but depending on how you calculate them, it may look like it is taking a longer time.

Regarding your questions:

> Why did bother the time to workerStart as well as swInit reduce? Or rather, what is captured as the time before workerStart and what is captured by service worker init exactly?

Just to be sure, could you tell us how the `workerStart` and `SWInit` are calculated? Generally, `workerStart` is recorded right before starting the SW and conducting fetch. Hence `workerStart` should contain operations such as redirects, and also the router evaluation time as well.

> Why did the response time improve as well? I would have expected the improved to come only from more efficient service worker handling

This is interesting. I think you are right, and response time should not change drastically. I guess it may seem that the response time is shorter depending on how you calculate it.
One scenario which this may happen is when you specify the race rule. For race rules, we start network requests before SW is started, so the response time would also be earlier. In addition, if the SW does things in the fetch handler other than sending network requests to fetch, then it would look if it took a longer time.
Let me check a bit more if there are other scenarios and get back to you.

**Comment by ErikWitt** [comment]
Hi @quasi-mod sorry for the delayed response but we implemented the tracking once again on a different page together with an A/B test for static routing so numbers are easier to compare.

These numbers show the breakdown of the median TTFB for returning user on Chrome mobile who landed on the homepage. The numbers are based on about 50 thousand page loads measured via real user monitoring. The static routing group simply has a network static routing rule for the homepage (that is the only difference between the groups).
![Image](https://github.com/user-attachments/assets/c30120be-6aa5-480a-9f01-f3d4e66a0daf)

This is how the metrics are computed based on the [navigation timing api](https://developer.mozilla.org/en-US/docs/Web/API/Performance_API/Navigation_timing) (if the timers are not set they are assumed to be 0)
* `redirectDuration = redirectEnd`
* `workerstart = workerStart - redirectDuration`
* `workerRouterEvaluationStart = workerRouterEvaluationStart - workerstart - redirectDuration`
* `swInit = fetch - workerRouterEvaluationStart - workerStart`
* `responseTime = finalResponseHeaderStart - fetch`

**Note:**  The data we see here likely contains cross origin redirects which are visible in the very high `workerstart` metric.

From the chart we can see that static routing improved:
* For a fair comparison, we should compare `workerstart + workerRouterEvaluationStart` to each other. Static routing improved the metric by 191 ms.
* `swInit` improved by 41 ms.
* Server response time improved by 57 ms.

What I do not quite understand is: 

What does `workerstart` in the control group contain? I guess it contains the otherwise invisible cross origin redirect time but what else is included in it? Also, what is included in the `swInit` as defined in this example?

**Comment by quasi-mod** [comment]
@ErikWitt 
Thanks for the context, and sorry for taking a long time to respond.
The data is indeed interesting, and we might need a bit more clarification.

> What does workerstart in the control group contain? I guess it contains the otherwise invisible cross origin redirect time but what else is included in it?

The workerStart resource timing field is specified to be right before starting the SW, so the workerStart in your description (RedirectEnd ~ WorkerStart) in the control group would include anything that happens before that. This shouldn't be long, so the data you are witnessing is a bit odd.

One possibility would be that when no redirect happens (in this case, the redirect end should be 0), hence it would be the equivalent to calculating from NavigationStart to the WorkerStart. Is there any chance this would be the case?

> Also, what is included in the swInit as defined in this example?

Looking at your calculation, it seems like SWInit is calculating workerStart ~ fetchStart, hence would include SW start time and calling the fetch handler. 

However, there might be a couple of clarification, especially for the treatment group: 
1) the router evaluation time typically happens before workerStart. I believe your calculation of `workerRouterEvaluationStart` is calculating the opposite?
2) when the router rule is network, we will not start the ServiceWorker, hence the workerStart should be zero (please refer to the demo site), could you double check this is the case?


Also, I wrote a doc to show couple of example use cases for the new fields to calculate what might be interesting:
https://docs.google.com/document/d/1vUMQtGIrPGTM8GfJWnnkVV2DI2S4rpZAVi4K2aE91kI/edit?tab=t.0

**Comment by ErikWitt** [comment]
@quasi-mod 

> One possibility would be that when no redirect happens (in this case, the redirect end should be 0), hence it would be the equivalent to calculating from NavigationStart to the WorkerStart. Is there any chance this would be the case?

Yes, for almost all of the traffic that went into this chart, `redirectEnd` is 0 and we are effectively looking at the time from navigation start. I suspect the reason that the time is so high is that the traffic contains lots of cross origin redirects (which are not visible in the timings.

Were it not for these cross origin redirects, I would expect `workerstart` to be almost 0 or is there any other processing happening between `navigationstart` and `workerstart` if there are no redirects and no static routing rules defined?

--- 

> the router evaluation time typically happens before workerStart. I believe your calculation of workerRouterEvaluationStart is calculating the opposite?
> when the router rule is network, we will not start the ServiceWorker, hence the workerStart should be zero (please refer to the demo site), could you double check this is the case?

Ah, you are right. In the treatment group, we route all these navigations to network so `workerStart` is always 0 in almost all cases (I double checked). This means the definition is equivalent to this in the treatment group.
```
workerRouterEvaluationStart = workerRouterEvaluationStart - redirectDuration
```

Thanks for the clarifications. I do now understand what `swInit` includes and it makes sense that it is reduced a lot.

What I am still confused by is this
* `workerstart` in the control group is 591ms (probably because it includes cross origin redirects). Since the redirects also happen in the treatment group, why is `workerRouterEvaluationStart` with 363ms so much faster. Shouldn't it include essentially the same time that was included in `workerstart` in the control group?

**Comment by ErikWitt** [comment]
Besides the timer discussion, `workerMatchedRouterSource` is extremely useful for us to attribute the source of the navigate request in our tracking.

Are you planning to release this in one of the next Chrome versions? It would be highly appreciated from our end
